pkg/grub/patches-2.06/0011-Add-measurefs-command.patch

```diff
@@ -0,0 +1,393 @@
++ *  GNU General Public License for more details.
++ *
++ *  You should have received a copy of the GNU General Public License
++ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
++ */
++
++#include <grub/device.h>
++#include <grub/fs.h>
++#include <grub/env.h>
++#include <grub/partition.h>
++#include <grub/i18n.h>
++#include <grub/extcmd.h>
++#include <grub/mm.h>
++
++#include <grub/crypto.h>
++#include <grub/tpm.h>
++
++GRUB_MOD_LICENSE ("GPLv3+");
++
++static const struct grub_arg_option options[] =
++  {
++    {"pcr", 'p', 0, N_("Select PCR register index to measure into. Default 9"), 0, ARG_TYPE_INT},    
++    {"hash", 'h', 0, N_("Hash type. Default sha256"), 0, ARG_TYPE_STRING},    
++    {0, 0, 0, 0, 0, 0}
++  };
++
++static grub_err_t
++grub_cmd_measurefs (grub_extcmd_context_t ctxt, int argc, char **args)
++{
++    struct grub_arg_list *state = ctxt->state;
++    grub_device_t dev;
++    grub_fs_t fs;
++    grub_err_t err = GRUB_ERR_NONE;
++
++    char* name = NULL;
++    int pcr = GRUB_BINARY_PCR;
++    char* hashalg = "sha256";
++    GRUB_PROPERLY_ALIGNED_ARRAY (result, GRUB_CRYPTO_MAX_MDLEN);
++    char* result_str = NULL;
++    int result_len;
++    int j;
++
++    if (argc > 0) {
++        name = args[0];
++    } else {
++        err = grub_error (GRUB_ERR_BAD_ARGUMENT, N_("device name expected"));
++        goto error_no_close;
++    }
++
++    // PCR index
++    if (state[0].set) {
++        pcr = grub_strtoul (state[0].arg, 0, 10);
++    }
++
++    // name of hash algorithm
++    if (state[1].set) {
++        hashalg = state[1].arg;
++    }
++
++    grub_printf("measurefs: Measuring %s into PCR-%d\n", name, pcr);
++
++    dev = grub_device_open(name);
++
++    if (!dev) {
++        err = grub_errno;
++        goto error_no_close;
++    }
++
++    if (dev->disk == NULL && dev->net != NULL) {
++        err = grub_error (GRUB_ERR_BAD_DEVICE,  N_("Network devices [`%s'] are not supported"), name);
++        goto error;
++    }
++
++    fs = grub_fs_probe (dev);
++
++    if (!fs) {
++        err = grub_error (GRUB_ERR_BAD_FS,  N_("cannot find a filesystem on `%s'"), name);
++        goto error;
++    }
++
++    grub_dprintf("measurefs", "FS: %s\n", fs->name);
++
++    if (fs->digest) {
++        err = fs->digest(dev, hashalg, &result, &result_len);
++
++        if (err == GRUB_ERR_NONE) {
++            // each byte is 2 chars + zero terminator
++            result_str = grub_malloc(result_len * 2 + 1);
++            if (result_str == NULL) {
++                err = GRUB_ERR_OUT_OF_MEMORY;
++                goto error;
++            }
++
++            for (j = 0; j < result_len; j++)
++                grub_snprintf(result_str + j * 2, 3, "%02x", ((grub_uint8_t *) result)[j]);
++            
++            char *desc = grub_xasprintf("%s %s", fs->name, result_str);
++            if (!desc)
++	        return GRUB_ERR_OUT_OF_MEMORY;
++
++            err = grub_tpm_measure(result, result_len, pcr, desc);
```

Comment:
Only this part has been changed, because grub 2.06 has its own grub_tpm_measure func.

in grub 2.06:
https://github.com/rhboot/grub2/blob/ae94b97be2b81b625d6af6654d3ed79078b50ff6/grub-core/commands/efi/tpm.c#L224-L241

in our grub:
https://github.com/lf-edge/eve/blob/aa23518d3d87b60920b50bb196627f0fa9058c81/pkg/grub/patches/0000-core-os-merge.patch#L2761-L2771

@mikem-zed is everything okay?